### PR TITLE
Fixed #419 (multiple document updates in single transaction)

### LINKED ISSUE #419: Allow a doc to be updated more than once in a single transaction
Should be ported from upstream ios: https://github.com/couchbase/couchbase-lite-ios/pull/589

### PR DESCRIPTION
**Pull request by hideki**
- Cache.java, DocumentChange.java and RevisionInternal.java - implemented missing methods
- Document.java - updated revisionAdded() method, and implemented missing method
- Database.java - Implemented storageExitedTransaction and databaseStorageChanged to notify document change during transaction, and also implemented missing method

### CODE REVIEW

**Comment by sergio91pt** [comment]
I'm not sure as I'm not familiar enough with the codebase to review this PR, but seems it exposes revisions not yet committed to other threads (MySQL transactions are thread local)

I made [this test](https://github.com/sergio91pt/couchbase-lite-android/commit/ae76cb61b3e2a48ab63383c955b5b0bd164f236a) and it fails on this branch but not on master.

File: src/main/java/com/couchbase/lite/Database.java [hunk 1]
```diff
@@ -3610,50 +3608,63 @@ private void postChangeNotifications() {
...
+            // TODO: change this to match iOS and call cachedDocumentWithID
+            //cachedDocumentWithID(change.getDocumentId()).revisionAdded(change, false);
+            getDocument(change.getDocumentId()).revisionAdded(change, false);
+        }
 
-    private void notifyChanges(List<DocumentChange> documentChanges) {
-        if (changesToNotify == null) {
-            changesToNotify = new ArrayList<DocumentChange>();
+        // Squish the change objects if too many of them are piling up:
+        if (changesToNotify.size() > MANY_CHANGES_TO_NOTIFY) {
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/Database.java]: Should it be >= ?

File: src/main/java/com/couchbase/lite/Cache.java [hunk 2]
```diff
@@ -76,6 +76,7 @@ public void clear() {
         weakReferenceCache.clear();
     }
 
-
-
+    public V resourceWithCacheKeyDontRecache(K key) {
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/Cache.java]: Are we using this anywhere or plan to use this?

**Comment by pasin** [comment]
I believed iOS has only one (only one level)  document cache. 

1. It's a weak smart cache that if a document inside a cache doesn't get retained by any ones (e.g. reference count = 0), the document will get automatically removed from the cache. Otherwise, it will stay inside the cache. So in normal case, if a document is still get used or referenced by something, getting the document from the cache should not be null.

2. Inside the cache itself, it also makes strong references to documents up to 50 (recently-used) docs so those 50 recently used docs get kept inside the cache even though they are not get retained by any other objects outside the cache.

3. When the database.purgeDocument: method is called, the purged document will get explicitly removed from the cache.

**Comment by hideki** [comment]
Hi @sergio91pt,
Thank you very much for reviewing the PR.
CBL Android uses SQLiteDatabase to access SQLite. SQLiteDatabase.beginTransaction() ( http://developer.android.com/reference/android/database/sqlite/SQLiteDatabase.html#beginTransaction() ) is a thread safe.

File: src/main/java/com/couchbase/lite/Database.java (hunk 1 above)
Comment (hideki) [inline src/main/java/com/couchbase/lite/Database.java]: @pasin Thanks a lot.

File: src/main/java/com/couchbase/lite/Cache.java (hunk 2 above)
Comment (hideki) [inline src/main/java/com/couchbase/lite/Cache.java]: This is equivalent implementation of `- (id<CBLCacheable>) resourceWithCacheKeyDontRecache: (NSString*)cacheKey;` in CBLCache.h/.m. I'd like to keep it.

**Pull request merged**